File: OlderCodes/Fortran77codes/ising.f
c  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
c  by RH Landau, MJ Paez, and CC BORDEIANU 
c  Copyright Princeton University Press, Princeton, 2008.
c  Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
c  MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
c  Support by National Science Foundation                             
c
c  ising.f: Ising model of magnetic dipole string  
c          If your compiler complains about drand48, seed48 
c          replace drand48 with rand(seed) and remove seed48 call 
c          Plot without connecting datapoints with lines

	    Program Ising
			
	      Implicit none
	      Integer max
	      Parameter(max = 100)
	      Integer element, i, spins(max), seed, t 
	      Real*8 drand48, energy, kt, new, j, old
c             define number temperature, exchange energy, random seed
	      Parameter(kt = 100, j = -1, seed = 68111)
c                                          open files, seed generator

	      Open(8, file = 'spin-up.dat', Status = 'Unknown')
	      Open(9, file = 'spin-do.dat', Status = 'Unknown')
	      Call seed48(seed)
c                           generate a uniform configuration of spins
	      Do 10 i = 1,max
          spins(i) = 1                     
 10     Continue
c                                                   step through time
	      Do 20 t = 1, 500
	        old = energy(spins, j, max)
	        element = drand48()*max+1
             
c                                                           flip spin
	        spins(element) = spins(element)*(-1)
	        new = energy(spins, j, max)
c                                                Metropolis algorithm
          If (new.GT.old .AND. exp((-new+old)/kt) .LT.drand48()) Then
            
             spins(element) = spins(element)*(-1)
          Endif
	        Do 30 i = 1,max
	        If (spins(i) .EQ. 1) Then
	          Write(8,*) t, i
	        Endif
	        If (spins(i) .EQ. (-1)) Then
	          Write(9,*) t, i
	        Endif
 30 	    Continue	 
 20     Continue
 	      Close(8)
 	      Close(9)
        Stop 'data saved in spin-up.dat, spin-do.dat'
 	    End 

	     Function energy(array, j, max)
	       Implicit none
	       Integer array(max), i, max
	       Real*8 energy, j
	       energy = 0
	       Do 22 i = 1,(max-1)
	         energy = energy+array(i)*array(i+1)
 22	     Continue
	       Return
	    End

File: OlderCodes/Ccodes/ising.c
/* From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
   Support by National Science Foundation                              
   */
//  ising.c: ising model of magnetic dipole string         
// Plot without conneting datapoints with lines  

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// if you don't have drand48 uncomment the following two lines 
    #define drand48 1.0/RAND_MAX*rand 
    #define srand48 srand        

#define max 100                                  // number of objects
#define kt 100.0                                       // temperature
#define J -1                                       // exchange energy
#define seed 68111                                // seed for srand48

main() {
	
  int i, j, element, array[max];
  double olden, newen;
  double energy(int array[]);                     // energy of system
	
  FILE *output1, *output2;                       // save spin ups and
  output1= fopen("spin-up.dat" ,"w");           // downs in two files
  output2= fopen("spin-down.dat", "w");
  srand48(seed);                                    // seed generator
	                                                   // uniform start
  for (i = 0; i<max; i++) array[i] = 1;   
	                                                       // time loop
  for (i = 0; i<= 500; i++)  {                         
    olden  = energy(array);                         // initial energy
    element= drand48()*max;                       // pick one element
    array[element] *= -1;                              // change spin
    newen  = energy(array);                   // calculate new energy
		                                          // reject/accept change
    if ( (newen>olden) && (exp((-newen + olden)/kt) <=  drand48()) )
      { array[element]= array[element]*(-1);   }  
		                                           // save "map" of spins
    for (j = 0; j<max; j++)   {               
      if (array[j] ==  1) fprintf(output1, "%d  %d\n", i, j);
      if (array[j] == -1) fprintf(output2, "%d  %d\n", i, j);
    }
  }
  fclose (output1);
  fclose (output2);
  printf("data saved in spin-up.dat, spin-down.dat\n");
}                                              // end of main program

double energy (int array[]) {              // function returns energy
	
  int i;
  double sum= 0.;  
  for (i = 0; i<(max-1); i++)                // loop through elements
     sum += array[i]*array[i+1];     
  return (J*sum);
}
